### ISSUE TITLE
Tag for spoilers

### ISSUE THREAD

**Issue by smelamud** [MoeraOrg/moera-issues#39]
Add a tag for spoilers: it hides its content under "Spoiler!" alert and shows only after clicking on it.

**Comment by eliduvid** [comment]
I think we can already just allow `<details>` html tag in the client. Than it's possible to make an extension to flexmark that will add it to markdown. Stack overflow uses for it blockquotes with `!` at every line, we can do it too.
I will be happy to do it. I think first part I can do even today.

**Comment by smelamud** [comment]
@eliduvid Cool! We can allow `<details>` and `<summary>` tags first, they seem very useful. Spoiler tag may be implemented then with a small CSS, like this https://jsfiddle.net/jeaw3upx/11/.

We even can create a custom HTML element for this. In this case user may simply use `<spoiler>` element in the markup and it will work! I never tried to implement custom elements, but it should be easy and very well fits our case.

I don't like blockquotes with `!` markup for this, because spoiler element should be inline, not block-level. But we may use `||spoiler text||` syntax that Discord uses.

If you can take this, it would be wonderful :wink:

**Comment by eliduvid** [comment]
@smelamud, I've seen many sites with block level spoilers (all phpbb, habr, SO) and it's useful for both for spoilers and for dumping large amount of data under it.

I like the idea of inline spoilers because SO syntax for block spoilers is _ugly_. Markdown was not meant for this sht.
On the other hand with block spoilers client side implementation of it will be little to none (default css for summary tag is OK for block spoilers).

I read about custom elements, it has some, er, properties:
* we can't use name `spoiler`, custom elements should have `-` in their name, like `<moera-spoiler>` (all simple names are reserved for future use). 
* it needs polifill for ie, early versions of edge and some others.

But generally writing it is pretty simple. 

On the other hand, we can implement it with react, I think you know better than me what can you do with it.

BTW, does moera support html tags inside markdown? This feature is part of CommonMark, but not everybody likes it.

**Comment by smelamud** [comment]
@eliduvid 

> @smelamud, I've seen many sites with block level spoilers (all phpbb, habr, SO) and it's useful for both for spoilers and for dumping large amount of data under it.

There is another similar proposal for collapsing large blocks of text. See #235. I wonder how we can implement both in a simplest way/

> I read about custom elements, it has some, er, properties:
> 
>     * we can't use name `spoiler`, custom elements should have `-` in their name, like `<moera-spoiler>` (all simple names are reserved for future use).

Hmm... OK, we can replace the tag on any stage: (a) immediately after converting Markdown to HTML; (b) after receiving the post body in the client from REST API call; (c) when inserting the post body into the page.

But, if we replace the `<spoiler>` tag anyway, we can replace it with `<details>/<summary>`, so the custom element is not needed.

>     * it needs polifill for ie, early versions of edge and some others.

I'm not too concerned about this. The general support is [pretty good](https://caniuse.com/#feat=custom-elementsv1), IE and pre-Chrome Edge will end their life in several months from now. In most cases I plan for the latest versions of Chrome & Firefox and maybe Safari.

> On the other hand, we can implement it with react, I think you know better than me what can you do with it.

Why to use React, if it works with simple HTML/CSS? And it should work in Web UI also.

> BTW, does moera support html tags inside markdown? This feature is part of CommonMark, but not everybody likes it.

Yes, it does. And I like it, because it gives us and users more freedom.

**Comment by eliduvid** [comment]
OK, there are:
1. Block spoiler. Behaves just like `details` tag: block, opens on click, collapsible back.
2. Inline spoiler. Inline, opens on click, not collapsible back.
3. Dummy inline spoiler. Just text with black background: users can see it if they select it.
4. Cut. Looks like "read more" link in the feed and like rest of the post in post view. I think it's usually implemented by some anchor or element that server is aware of and two different rest endpoints for full post and summary of it. 
5. Dummy cut. Facebook approach to cut. Automatically inserted after some amount of characters. 

There are some use-cases:
1. Spoiler hiding. As I have seen people are using just anything to do it, even a dummy cut.
Generally any kind of spoiler implementation is OK, as long as it can be _multiple_ of them in the same post. Inline spoilers are good for hiding parts of the sentence, but many places just don't have those, and I've not seen anybody get mad about needing to use block spoilers instead.
2. Collapsing large blocks of text inside the post. Classic use for that are technical posts with chunks of code or logs throughout the text. Like: "I did this and got this (logs), than I tried some other things and got this (logs). Here's the full code (code)".
I've seen block spoilers used for it and nothing else, because there should be multiple of those, and they must be collapsible back.
3. Short summary in feed instead of full posts. Best way to do it are cuts. Dummy cuts are bad, block spoilers IMO are OK for now, but people need cuts.

So, it looks for me that we need cuts, block spoilers and maybe inline spoilers or dummy inline spoilers. I'd leave cuts for now (there's a separate issue for that).
Block spoilers generally cover more cases, but any markdown for them is ugly. There's already many blocks in markdown and only so much weird symbols we can use for it 😺
So maybe it's better to use only html for block spoilers, if it's not very complicated. And I don't think `details` tag is complicated.
